### PR TITLE
[Enhance] Support ZeRO3

### PR DESCRIPTION
**Pull request by yhna940**
## Title

- [Enhance] Support ZeRO3

## Description

This PR introduces an implementation of ZeRO3, motivated by `ColossalAI`. ZeRO (ZeRO-Redundancy Optimizer) is a novel memory optimization technology developed by Microsoft. ZeRO3 is designed to train large models by partitioning model parameters, gradients, and optimizer states across data parallel processes.

The implementation primarily consists of four key components:

**Model Wrapper:** This component scatters and gathers parameters as needed throughout the training process, making the model compatible with the sharding and reassembly of parameters that ZeRO3 entails.

**Optimizer Wrapper:** This component is responsible for updating the sharded parameters. It adapts the underlying optimizer to operate on sharded model parameters, gradients, and optimizer states.

**Hooks:** These record the training phase and gather the parameters accordingly. The hooks are necessary to manage the lifecycle of parameters, ensuring that they exist in the right place at the right time.

**Chunks:** The purpose of chunks is to manage flattened tensors to enhance communication efficiency. This is an important feature for managing memory usage, especially for large models where memory overhead can be a challenge.

Please review the following changes. Your valuable feedback and comments are much appreciated.

## Linked Issues

- N/A

## Usage

ZeRO3 can be used as follows in your training script:

```python
from oslo.torch.nn.parallel.data_parallel.zero import (
    _FullyShardedDataParallel,
    _HeterogeneousZeroOptimizer,
)

# Initialize your model
model = ...  # your model definition here

# Wrap the model with FullyShardedDataParallel
model = _FullyShardedDataParallel(model, device, parallel_context=parallel_context)
model.parallelize()

# Define an optimizer and wrap it with HeterogeneousZeroOptimizer
base_optimizer = ...  # your optimizer definition here
optimizer = _HeterogeneousZeroOptimizer(base_optimizer, model)

# Typical training step
optimizer.zero_grad()
inputs = inputs.to(device)
labels = labels.to(device)
outputs = model(inputs)
loss = calc_loss(outputs, labels)
loss.backward()
optimizer.step()
```

This is a typical usage example. Please adapt the usage according to your specific training code.

### CODE REVIEW

**Comment by yhna940** [comment]
# Experiment

We have conducted a series of experiments to evaluate our ZeRO3 implementation.

## Convergence Evaluation

We used the CIFAR10 dataset and a ResNet-50 backbone for evaluating the convergence of our implementation. Our results confirm that our ZeRO3 implementation converges without any issues.

Epoch 7: validation accuracy = 0.9482

```python
import logging
import os
import torch
import torch.nn as nn
import torch.optim as optim
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.utils.data import DataLoader, DistributedSampler
from torchvision import datasets, transforms, models

from oslo.torch.distributed.parallel_context import ParallelContext
from oslo.torch.nn.parallel.data_parallel.zero import _FullyShardedDataParallel
from oslo.torch.nn.parallel.data_parallel.zero.sharded_optim.heterogeneous_optim import (
    _HeterogeneousZeroOptimizer,
)

def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12345'
    os.environ["RANK"] = str(rank)
    os.environ["LOCAL_RANK"] = str(rank)
    os.environ["WORLD_SIZE"] = str(world_size)
    os.environ["LOCAL_WORLD_SIZE"] = str(world_size)
    return ParallelContext.from_torch(data_parallel_size=world_size)

# Set up logger
class DistLogger:
    def __init__(self, log_file_path):
        self.logger = logging.getLogger(__name__)
        self.logger.setLevel(logging.DEBUG)
        formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
        file_handler = logging.FileHandler(log_file_path)
        file_handler.setLevel(logging.DEBUG)
        file_handler.setFormatter(formatter)
        self.logger.addHandler(file_handler)

    def debug(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.debug(msg, *args, **kwargs)

    def info(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.info(msg, *args, **kwargs)

    def warning(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.warning(msg, *args, **kwargs)

    def error(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.error(msg, *args, **kwargs)

    def critical(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.critical(msg, *args, **kwargs)

logger = DistLogger('zero-res-training.log')

# Define the data transformation
transform = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])

# Load the CIFAR10 dataset
train_dataset = datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
val_dataset = datasets.CIFAR10(root='./data', train=False, download=True, transform=transform)

# Set up distributed training
def train(rank, world_size):
    parallel_context = setup(rank, world_size)
    dist.barrier()

    # Set up the model
    device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
    model = models.resnet50(weights=models.resnet.ResNet50_Weights.IMAGENET1K_V2)
    model.fc = nn.Linear(2048, 10) # CIFAR10 has 10 classes
    if torch.cuda.is_available() and torch.cuda.device_count() > 1:
        model = _FullyShardedDataParallel(model, device, parallel_context=parallel_context)
        model.parallelize()
    else:
        raise Exception
    # Set up optimizer
    optimizer = optim.AdamW(model.parameters(), lr=1e-4)
    optimizer = _HeterogeneousZeroOptimizer(optimizer, model)

    if torch.cuda.is_available() and torch.cuda.device_count() > 1:
        train_sampler = DistributedSampler(train_dataset, num_replicas=world_size, rank=rank)
        train_loader = DataLoader(train_dataset, num_workers=4, batch_size=16, sampler=train_sampler)
        val_loader = DataLoader(val_dataset, num_workers=4, batch_size=64)
    else:
        raise Exception

    num_epochs = 8
    for epoch in range(num_epochs):
        # Training loop
        model.train()
        for idx, (inputs, labels) in enumerate(train_loader):
            optimizer.zero_grad()
            inputs = inputs.to(device)
            labels = labels.to(device)
            outputs = model(inputs)
            loss = nn.CrossEntropyLoss()(outputs, labels)
            loss.backward()
            optimizer.step()
            if idx % 10 == 0:
                logger.info(f'Epoch {epoch} Iter {idx}: loss = {loss.item():.4f}, memory = {torch.cuda.memory_allocated() / 1024 / 1024:.4f}')

        # Evaluation loop
        model.eval()
        correct = 0
        total = 0
        for inputs, labels in val_loader:
            inputs = inputs.to(device)
            labels = labels.to(device)
            outputs = model(inputs)
            predictions = outputs.argmax(dim=-1)
            correct += (predictions == labels).sum().item()
            total += labels.shape[0]
        accuracy = correct / total
        logger.info(f'Epoch {epoch}: validation accuracy = {accuracy:.4f}')


if __name__ == '__main__':
    world_size = 4
    mp.spawn(train, args=(world_size,), nprocs=world_size, join=True)
    
```

## Memory Benchmark

We benchmarked memory usage of our ZeRO3 implementation in comparison to standard DistributedDataParallel.

Here are our results:

| Technique                               | Peak memory (MiB) |
| ---------------------------- | --------------------|
| ZeRO3                                     | 522.9                        |
| DistributedDataParallel          | 917.2                         |

These results indicate that our ZeRO3 implementation significantly reduces memory footprint compared to traditional DistributedDataParallel.

```python
import logging
import os
import torch
import torch.nn as nn
import torch.optim as optim
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.utils.data import DataLoader, DistributedSampler
from torchvision import datasets, transforms, models

from oslo.torch.distributed.parallel_context import ParallelContext
from oslo.torch.nn.parallel.data_parallel.zero import _FullyShardedDataParallel
from oslo.torch.nn.parallel.data_parallel.data_parallel import _DistributedDataParallel
from oslo.torch.nn.parallel.data_parallel.zero.sharded_optim.heterogeneous_optim import (
    _HeterogeneousZeroOptimizer,
)

def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12345'
    os.environ["RANK"] = str(rank)
    os.environ["LOCAL_RANK"] = str(rank)
    os.environ["WORLD_SIZE"] = str(world_size)
    os.environ["LOCAL_WORLD_SIZE"] = str(world_size)
    return ParallelContext.from_torch(data_parallel_size=world_size)

# Set up logger
class DistLogger:
    def __init__(self, log_file_path):
        self.logger = logging.getLogger(__name__)
        self.logger.setLevel(logging.DEBUG)
        formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
        file_handler = logging.FileHandler(log_file_path)
        file_handler.setLevel(logging.DEBUG)
        file_handler.setFormatter(formatter)
        self.logger.addHandler(file_handler)

    def debug(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.debug(msg, *args, **kwargs)

    def info(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.info(msg, *args, **kwargs)

    def warning(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.warning(msg, *args, **kwargs)
                                                                                                                                                                                        3,1           Top
            model.parallelize()

    optimizer = optim.AdamW(model.parameters(), lr=1e-4)

    if optimizer_class is not None:
        optimizer = optimizer_class(optimizer, model)

    dummy_input = torch.randn(4, 3, 224, 224, device=device)
    dummy_target = torch.randint(0, 100, (4,), device=device)

    criterion = nn.CrossEntropyLoss()

    optimizer.zero_grad()
    outputs = model(dummy_input)
    loss = criterion(outputs, dummy_target)
    loss.backward()
    optimizer.step()

    peak_memory = torch.cuda.max_memory_allocated(device)

    # Cleanup
    del model
    del optimizer
    del loss
    del outputs
    torch.cuda.empty_cache()

    return peak_memory

# Set up distributed training
def main(rank, world_size):
    parallel_context = setup(rank, world_size)
    dist.barrier()

    # Set up the model
    device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')

    # Measure memory with FullyShardedDataParallel and HeterogeneousZeroOptimizer
    peak_memory_1 = measure_memory(device, _FullyShardedDataParallel, _HeterogeneousZeroOptimizer, parallel_context)
    logger.info(f'Peak memory with FullyShardedDataParallel and HeterogeneousZeroOptimizer: {peak_memory_1}')

    # Measure memory with DistributedDataParallel
    peak_memory_2 = measure_memory(device, _DistributedDataParallel, None, parallel_context)
    logger.info(f'Peak memory with DistributedDataParallel: {peak_memory_2}')

if __name__ == '__main__':
    world_size = 2
    mp.spawn(main, args=(world_size,), nprocs=world_size, join=True)
```
